Singleton class

rajeevgUCI opened the pull request:
Created Singleton class and refactored GameManager to use it.

Testing: wrote tests for Singleton class. Removed GameManagerTests that were only testing singleton functionality.

rajeevgUCI committed:
Add SingletonMonoBehavior class

rajeevgUCI committed:
Add Singleton tests

rajeevgUCI committed:
Refactor GameManager to use SingletonMonoBehavior

Refactor GameManager to inherit from SingletonMonoBehavior. Also remove
Singleton testing from GameManagerTests.

rajeevgUCI committed:
Fix SingletonMonoBehavior code style

Change code style to match rest of project. Write better log messages
and comments.

rajeevgUCI committed:
SingletonTestComponent constructor -> protected

rajeevgUCI committed:
Remove costly check for multiple singletons

rajeevgUCI committed:
Change new singleton instance message

Also minor refactoring.

rajeevgUCI commented:
(Rewrite still in progress, do not review yet.)

rajeevgUCI committed:
SingletonMonoBehavior: destroy extraneous instance

rajeevgUCI committed:
Rewrite SingletonTests, note destroy difficulties

rajeevgUCI committed:
Set singleton Instance on Awake, for performance

rajeevgUCI committed:
Add singleton tests for first Instance and destroy

kylebartz commented on Assets/Runtime/core/GameManager.cs:
```diff
@@ -1,18 +1,11 @@
-    } else if (manager != this) {
-      Destroy(this.gameObject); // Delete current instance to maintain singleton.
-    }
+    this.Initialize();
```
The null check safety around Initialize is necessary here. GameManager's Initialize should only ever be called once no matter what. Without a safety check, in the off chance that there are multiple GameManagers in the scene there will be unintended effects (such as double bindings being registered). You can quickly test this by adding two game managers on the scene on master branch vs this branch.

rajeevgUCI replied:
Good catch - will fix.

rajeevgUCI replied:
Fixed.

kylebartz commented on Assets/Runtime/core/GameManager.cs:
```diff
@@ -1,18 +1,11 @@
+public class GameManager : SingletonMonoBehavior<GameManager> {
   private GameState state = GameState.ACTIVE;
 
   private void Awake() {
```
This Awake method is overriding the inherited `Awake()` from `SingletonMonoBehavior`. That means your instance check and deleting if duplicate will never run. It is then possible for multiple instances of GameManger to exist and be active in the scene. Only the first will be registered as the singleton instance, but the rest will still be active and ticking on Update

One solution would be to set `SingletonMonoBehavior`'s `Awake` method to `virtual`, and then have `GameManager`'s `override` and call the base Awake.

rajeevgUCI replied:
I missed that Unity private methods inherit, unlike normal C# private methods. Will fix per your suggestion, then make sure Unity won't e.g. ignore the override and call both Awake methods.

rajeevgUCI replied:
Fixed. Also similarly fixed OnDestroy.

kylebartz commented on Assets/Tests/Runtime/core/SingletonTests.cs:
```diff
@@ -0,0 +1,49 @@
+using NUnit.Framework;
+using System.Collections;
+
+public class SingletonTests {
```
This is way over commented to the point where it is hard to read the code. In the future try to keep comments minimal.

rajeevgUCI replied:
The comments document subtleties not apparent from reading the code. So I think it's best to keep their content, but I'll try to make them more concise.

rajeevgUCI committed:
SingletonTests: add first Instance null test

rajeevgUCI committed:
SingletonTests: make comments more concise

rajeevgUCI committed:
Shorten line length in SingletonTests

rajeevgUCI committed:
Do not update null instance, instead only Assert

rajeevgUCI committed:
Set Unity methods to protected, to prevent hiding

Fix bug where subclass could prevent SingletonMonoBehavior Unity methods
(Awake, OnDestroy) from being called by redefining them. Set those
methods to protected, and document that the base class methods should be
called.

rajeevgUCI committed:
Add Awake and OnDestroy overrides to testing

Test that overrides do not break SingletonTestComponent functionality.

rajeevgUCI committed:
Update GameManager to correctly subclass singleton

Update GameManager to correctly subclass SingletonMonoBehavior, per the
comments in SingletonMonoBehavior.

rajeevgUCI committed:
Fix GameManager to call Bindings.Initialize once

rajeevgUCI commented:
Made changes per feedback above. Currently resolving merge conflicts.

rajeevgUCI committed:
Merge branch 'master' into singleton-class

kylebartz committed:
Added a DontDestroyOnLoad to Awake for when Awake registers before the Property

kylebartz committed:
Set Awake's Destroy to delete the game object (will delete everything attached)
